flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java

```diff
@@ -126,9 +137,16 @@ public void initializeState(StateInitializationContext context) throws Exception
     this.manifestOutputFileFactory = FlinkManifestUtil.createOutputFileFactory(table, flinkJobId, subTaskId, attemptId);
     this.maxCommittedCheckpointId = INITIAL_CHECKPOINT_ID;
 
+    Map<String, String> properties = this.table.properties();
+    this.currentWatermark = PropertyUtil.propertyAsLong(properties, CURRENT_WATERMARK, -1L);
+    this.storeWatermark = PropertyUtil.propertyAsBoolean(properties, STORE_WATERMARK, false);
+
     this.checkpointsState = context.getOperatorStateStore().getListState(STATE_DESCRIPTOR);
     this.jobIdState = context.getOperatorStateStore().getListState(JOB_ID_DESCRIPTOR);
+    this.watermarkState = context.getOperatorStateStore().getListState(WATERMARK_DESCRIPTOR);
+
     if (context.isRestored()) {
+      watermarkPerCheckpoint.putAll(watermarkState.get().iterator().next());
```

Comment:
Will this be backwards compatible for on going streaming jobs that don't have any `watermarkState` when they restore? For example, for on going streaming jobs that are upgraded to a version of iceberg that includes this patch?

Looking at the Flink `AppendingState` interface, it says that calling `.get()` should return `null` if the state is empty. Also, you can see that the value of `restoredFlinkJobId` below from calling `jobIdState.get().iterator().next()` below is checked for null or empty.

Reply:
Actually, on further inspection of the `ListState` interface, it says that passing `null` to `putAll` is a no-op. So I don' think there should be backwards compatibility issues, but should we possibly be (1) logging something if no watermark state is restored and/or (2) possibly doing our own `null` check vs relying on the documented behavior of `ListState#putAll` to be consistent over time when inserting `null`?

Reply:
Thanks @kbendick, i neglected backwards compatible thing, current code will raise an `java.util.NoSuchElementException` when `watermarkState` is empty,  i will fix it, also i will logging whether watermark state is restored since the restore is not an high frequency event.
BTW, `watermarkPerCheckpoint` is an instance of `HashMap`, i think the variable name misled you here :smile:.